Remove overlapping parts of multipart suggestions

lukas-code opened the pull request:
This PR adds a debug assertion that the parts of a single substitution cannot overlap, fixes a overlapping substitution from the testsuite, and fixes https://github.com/rust-lang/rust/issues/106870.

Note that a single suggestion can still have multiple overlapping substitutions / possible edits, we just don't suggest overlapping replacements in a single edit anymore.

I've also included a fix for an unrelated bug where rustfix for `explicit_outlives_requirements` would produce multiple trailing commas for a where clause.

Commit:
add debug assertion for suggestions with overlapping parts

Commit:
preserve delim spans during `macro_rules!` expansion if able

Commit:
fix overlapping spans for `explicit_outlives_requirements` in macros

also delete trailing comma if necessary

WaffleLapkin commented on compiler/rustc_errors/src/diagnostic.rs:
```diff
@@ -630,18 +630,27 @@ impl Diagnostic {
+            .map(|(span, snippet)| SubstitutionPart { snippet, span })
+            .collect::<Vec<_>>();
+
+        for (index, part_outer) in parts.iter().enumerate() {
+            debug_assert!(
+                !part_outer.span.is_empty() || !part_outer.snippet.is_empty(),
+                "Span must not be empty and have no suggestion",
+            );
+            for part_inner in parts.iter().skip(index + 1) {
+                debug_assert!(
+                    !part_outer.span.overlaps(part_inner.span),
+                    "suggestion must not have overlapping parts:\n{part_outer:?}\n{part_inner:?}",
+                );
+            }
+        }
```
Does it make sense to sort parts by the `span`? this way you only need to check the two adjacent ones. But, we shouldn't do this just for the debug assertions maybe...

Either way, I think it would be better to enclose this in `if cfg!(debug_assertions) {}`, so llvm has an easier time removing the useless loops.

Also, can we somehow test this assertion?

estebank replied:
I think it is perfectly safe to always unconditionally sort here.

WaffleLapkin commented:
r? @estebank

estebank commented on compiler/rustc_errors/src/diagnostic.rs:
```diff
@@ -630,18 +630,27 @@ impl Diagnostic {
+            .map(|(span, snippet)| SubstitutionPart { snippet, span })
+            .collect::<Vec<_>>();
+
+        for (index, part_outer) in parts.iter().enumerate() {
+            debug_assert!(
+                !part_outer.span.is_empty() || !part_outer.snippet.is_empty(),
+                "Span must not be empty and have no suggestion",
+            );
+            for part_inner in parts.iter().skip(index + 1) {
+                debug_assert!(
+                    !part_outer.span.overlaps(part_inner.span),
+                    "suggestion must not have overlapping parts:\n{part_outer:?}\n{part_inner:?}",
+                );
+            }
+        }
```
After sorting suggestion:

```suggestion
        let mut parts = parts.iter.peekable();
        while let Some(part) = parts.next() {
            debug_assert!(
                !partspan.is_empty() || !part.snippet.is_empty(),
                "Span must not be empty and have no suggestion",
            );
            let Some(next) = parts.peek() else { break; }
            debug_assert!(
                !part.span.overlaps(next.span),
                "suggestion must not have overlapping parts:\n{part:?}\n{next:?}",
            );
        }
```

WaffleLapkin replied:
I think this can be written nicer with `parts.array_windows()`. Something like
```rust
debug_assert!(parts.iter().all(|part| !part.span.is_empty() || !part.snipper.is_empty()), "Span must not be empty and have no suggestion");
parts.array_windows(|[a, b]| debug_assert!(!a.span.overlaps(b.span), "suggestion must not have overlapping parts:\n{a:?}\n{b:?}"));
```

estebank commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -224,12 +223,25 @@ fn expand_macro<'cx>(
+                    let rhs_tt = &rhs.tts[i];
+                    let ctxt = tt.span().ctxt();
+                    match (&mut tt, rhs_tt) {
+                        // preserve the delim spans if able
```
Nice

estebank commented:
r=me after addressing the sorting/assertion nitpick

WaffleLapkin commented:
@bors r=estebank

bors commented:
:pushpin: Commit d18768ae85a78def4d45affe1b12ce8b3d9aafc7 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank d18768ae85a78def4d45affe1b12ce8b3d9aafc7 -->
<!-- homu: {"type":"Approved","sha":"d18768ae85a78def4d45affe1b12ce8b3d9aafc7","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

lukas-code commented:
For the record, this is how this assertion looks now:
```text
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `None`,
 right: `Some([SubstitutionPart { span: /home/lukas/code/rust/tests/ui/parser/issues/issue-44406.rs:3:12: 3:24 (#4), snippet: " { " }, SubstitutionPart { span: /home/lukas/code/rust/tests/ui/parser/issues/issue-44406.rs:3:12: 3:24 (#4), snippet: " }" }])`: suggestion must not have overlapping parts', /home/lukas/code/rust/compiler/rustc_errors/src/diagnostic.rs:646:9
```

matthiaskrgr commented:
Might break clippy tests

WaffleLapkin commented:
It does look like it breaks clippy tests (diff from #107020):
```diff
-error: variables can be used directly in the `format!` string
-   |
-   |
-LL |     println!("{0} {0}", local_i32);
-   |
-help: change this to
-   |
-   |
-LL -     println!("{0} {0}", local_i32);
-LL +     println!("{local_i32} {local_i32}");
+thread 'rustc' panicked at 'assertion failed: `(left == right)`
+  left: `None`,
+  left: `None`,
+ right: `Some([SubstitutionPart { span: $DIR/uninlined_format_args.rs:78:23: 78:34 (#0), snippet: "" }, SubstitutionPart { span: $DIR/uninlined_format_args.rs:78:23: 78:34 (#0), snippet: "" }])`: suggestion must not have overlapping parts', /checkout/compiler/rustc_errors/src/diagnostic.rs:646:9
```

Not sure why clippy removes the same span twice though 🤔

WaffleLapkin commented:
@bors r-

Commit:
fix overlapping spans for `clippy::uninlined_format_args`

lukas-code commented:
I fixed the clippy failure and also changed the signature of `multipart_suggestion_*` to take `impl IntoIterator` instead of `Vec`, because I didn't like the `.into_iter().collect().into_iter().collect()`. I hope this doesn't affect compile time too much.

@rustbot ready

WaffleLapkin commented on compiler/rustc_hir_analysis/src/check/wfcheck.rs:
```diff
@@ -825,7 +825,8 @@ fn check_object_unsafe_self_trait_by_name(tcx: TyCtxt<'_>, item: &hir::TraitItem
             return;
         }
-        let sugg = trait_should_be_self.iter().map(|span| (*span, "Self".to_string())).collect();
+        let sugg =
+            trait_should_be_self.iter().map(|span| (*span, "Self".to_string())).collect::<Vec<_>>();
```
Wouldn't it make more sense to remove the collect altogether?

WaffleLapkin commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -2223,6 +2241,8 @@ impl<'tcx> LateLintPass<'tcx> for ExplicitOutlivesRequirements {
                     Applicability::MaybeIncorrect
                 };
 
+                let lint_spans = lint_spans.into_iter().collect::<Vec<_>>();
+
                 cx.emit_spanned_lint(
                     EXPLICIT_OUTLIVES_REQUIREMENTS,
                     lint_spans.clone(),
```
Remove `collect`? also the clone seems unnecessary? Can't this just be 
```suggestion
                cx.emit_spanned_lint(
                    EXPLICIT_OUTLIVES_REQUIREMENTS,
                    lint_spans,
```

lukas-code replied:
The clone is necessary, because we need one vector for the `MultiSpan` of the lint itself and one for the suggestion.

The collect is turning the `HashSet` into a `Vec` for the `MultiSpan`. But it would probably be better to just sort and dedup a vector instead of using a hash set, will change that.

WaffleLapkin commented on compiler/rustc_errors/src/diagnostic.rs:
```diff
@@ -625,23 +625,31 @@ impl Diagnostic {
-                    .map(|(span, snippet)| SubstitutionPart { snippet, span })
-                    .collect(),
-            }],
+            substitutions: vec![Substitution { parts }],
```
💭  would small vec be good here?

WaffleLapkin reviewed (commented):
Generally it's better to make smaller PR and separate refactorings, but this looks good overall, except for a couple of nitpicks.

WaffleLapkin commented:
@bors try @rust-timer queue
Let's see if additional generics made the compiler compilation slower.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 17a9aa973c87c2152ddcf30ccff76ae041c92b04 with merge 44d89d81a90f3329a8b60eb1367e28c97090ff24...
<!-- homu: {"type":"TryBuildStarted","head_sha":"17a9aa973c87c2152ddcf30ccff76ae041c92b04","merge_sha":"44d89d81a90f3329a8b60eb1367e28c97090ff24"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3968596556/jobs/6804774775)
Build commit: 44d89d81a90f3329a8b60eb1367e28c97090ff24 (`44d89d81a90f3329a8b60eb1367e28c97090ff24`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3968596556/jobs/6804774775"},"merge_sha":"44d89d81a90f3329a8b60eb1367e28c97090ff24"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3968596556/jobs/6804774775)
Build commit: 44d89d81a90f3329a8b60eb1367e28c97090ff24 (`44d89d81a90f3329a8b60eb1367e28c97090ff24`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3968596556/jobs/6804774775"},"merge_sha":"44d89d81a90f3329a8b60eb1367e28c97090ff24"} -->

rust-timer commented:
Queued 44d89d81a90f3329a8b60eb1367e28c97090ff24 with parent 56ee85274e5a3a4dda92f3bf73d1664c74ff9c15, future [comparison URL](https://perf.rust-lang.org/compare.html?start=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&end=44d89d81a90f3329a8b60eb1367e28c97090ff24).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (44d89d81a90f3329a8b60eb1367e28c97090ff24): [comparison URL](https://perf.rust-lang.org/compare.html?start=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&end=44d89d81a90f3329a8b60eb1367e28c97090ff24&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&end=44d89d81a90f3329a8b60eb1367e28c97090ff24&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.2% | [0.2%, 2.4%] | 11    |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.2%, 0.5%] | 9     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.2% | [0.2%, 2.4%] | 11    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&end=44d89d81a90f3329a8b60eb1367e28c97090ff24&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 5.7%  | [3.3%, 8.1%]   | 2     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.1%, -1.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-3.7%, -1.7%] | 3     |
| All ❌✅ (primary)                 | -1.1% | [-1.1%, -1.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=56ee85274e5a3a4dda92f3bf73d1664c74ff9c15&end=44d89d81a90f3329a8b60eb1367e28c97090ff24&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [1.1%, 1.5%]   | 6     |
| Regressions ❌ <br /> (secondary)  | 25.6% | [25.6%, 25.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.3%  | [1.1%, 1.5%]   | 6     |
</details>

WaffleLapkin commented:
I assume you've removed the generic changes, so let's see what perf says @bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 228ddf04fca01e8224ad4fb8a8e9db8483642249 with merge 5bfca626acb506f9da84b4eac21a87e0847f8785...
<!-- homu: {"type":"TryBuildStarted","head_sha":"228ddf04fca01e8224ad4fb8a8e9db8483642249","merge_sha":"5bfca626acb506f9da84b4eac21a87e0847f8785"} -->

WaffleLapkin commented:
@bors try

bors commented:
:hourglass: Trying commit 228ddf04fca01e8224ad4fb8a8e9db8483642249 with merge cddc29a75100f2566a46e957e0b0df708bc12c68...
<!-- homu: {"type":"TryBuildStarted","head_sha":"228ddf04fca01e8224ad4fb8a8e9db8483642249","merge_sha":"cddc29a75100f2566a46e957e0b0df708bc12c68"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4023626985/jobs/6916682246)
Build commit: cddc29a75100f2566a46e957e0b0df708bc12c68 (`cddc29a75100f2566a46e957e0b0df708bc12c68`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4023626985/jobs/6916682246"},"merge_sha":"cddc29a75100f2566a46e957e0b0df708bc12c68"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4023626985/jobs/6916682246)
Build commit: cddc29a75100f2566a46e957e0b0df708bc12c68 (`cddc29a75100f2566a46e957e0b0df708bc12c68`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4023626985/jobs/6916682246"},"merge_sha":"cddc29a75100f2566a46e957e0b0df708bc12c68"} -->

rust-timer commented:
Queued cddc29a75100f2566a46e957e0b0df708bc12c68 with parent 18890f05f6cea40fd25bb4bb9aa6f7372b69f641, future [comparison URL](https://perf.rust-lang.org/compare.html?start=18890f05f6cea40fd25bb4bb9aa6f7372b69f641&end=cddc29a75100f2566a46e957e0b0df708bc12c68).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (cddc29a75100f2566a46e957e0b0df708bc12c68): [comparison URL](https://perf.rust-lang.org/compare.html?start=18890f05f6cea40fd25bb4bb9aa6f7372b69f641&end=cddc29a75100f2566a46e957e0b0df708bc12c68&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=18890f05f6cea40fd25bb4bb9aa6f7372b69f641&end=cddc29a75100f2566a46e957e0b0df708bc12c68&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=18890f05f6cea40fd25bb4bb9aa6f7372b69f641&end=cddc29a75100f2566a46e957e0b0df708bc12c68&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.7% | [2.7%, 2.7%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.7% | [2.4%, 2.9%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.7% | [2.7%, 2.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=18890f05f6cea40fd25bb4bb9aa6f7372b69f641&end=cddc29a75100f2566a46e957e0b0df708bc12c68&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

lukas-code commented:
Regression is from the improvements in https://github.com/rust-lang/rust/pull/107343#issuecomment-1406299672, so probably noise.

WaffleLapkin commented:
@bors r=estebank

bors commented:
:pushpin: Commit 228ddf04fca01e8224ad4fb8a8e9db8483642249 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 228ddf04fca01e8224ad4fb8a8e9db8483642249 -->
<!-- homu: {"type":"Approved","sha":"228ddf04fca01e8224ad4fb8a8e9db8483642249","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 228ddf04fca01e8224ad4fb8a8e9db8483642249 with merge d8da51366817317d19be9bd138943eafb30bc987...
<!-- homu: {"type":"BuildStarted","head_sha":"228ddf04fca01e8224ad4fb8a8e9db8483642249","merge_sha":"d8da51366817317d19be9bd138943eafb30bc987"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4031269796/jobs/6931548966)
Approved by: estebank
Pushing d8da51366817317d19be9bd138943eafb30bc987 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"estebank","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4031269796/jobs/6931548966"},"merge_sha":"d8da51366817317d19be9bd138943eafb30bc987"} -->

rust-timer commented:
Finished benchmarking commit (d8da51366817317d19be9bd138943eafb30bc987): [comparison URL](https://perf.rust-lang.org/compare.html?start=226b2496fc3ef997fa1cccf84259539203f82375&end=d8da51366817317d19be9bd138943eafb30bc987&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=226b2496fc3ef997fa1cccf84259539203f82375&end=d8da51366817317d19be9bd138943eafb30bc987&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-1.2%, -0.1%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=226b2496fc3ef997fa1cccf84259539203f82375&end=d8da51366817317d19be9bd138943eafb30bc987&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.7% | [2.6%, 2.8%] | 2     |
| Regressions ❌ <br /> (secondary)  | 1.5% | [1.1%, 2.2%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.7% | [2.6%, 2.8%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=226b2496fc3ef997fa1cccf84259539203f82375&end=d8da51366817317d19be9bd138943eafb30bc987&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

WaffleLapkin commented:
@lukas-code do you still have the commit/diff of making diagnostic functions accept `impl Interator`? I would like to experiment with this a bit

lukas-code commented:
@WaffleLapkin I pulled this out of my reflog: 36d99530e2758c0099102bb9ae8e5d5cbc0bff2b https://github.com/lukas-code/rust/tree/multipart-impl-iterator

WaffleLapkin commented:
@lukas-code thanks!
